File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {

    
    int max = 0;



    int l = 0;
    int r = heights.size() -1;


    while ( l < r){

        int sum = 0;
        if( heights[l] >= heights[r]){

        sum = heights[r] * (r-l);
        r--;

        }

        else if( heights[l] <= heights[r]){
            sum = heights[l] * (r-l);
            l++;

        }


        if( sum > max) max = sum;

    }


    return max;

        
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        
 
        case1 = self.rob_linear(nums[:-1])
        
    
        case2 = self.rob_linear(nums[1:])
        
        return max(case1, case2)
    
    def rob_linear(self, nums: List[int]) -> int:
      
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        
        mem = [0] * len(nums)
        mem[0] = nums[0]
        mem[1] = max(nums[0], nums[1])
        
        for i in range(2, len(nums)):
            mem[i] = max(mem[i-1], mem[i-2] + nums[i])
        
        return mem[len(nums) - 1]